hal/src/electron/modem/mdm_hal.cpp

```diff
@@ -1271,85 +1323,62 @@ bool MDMParser::powerOff(void)
-                // Skip power off sequence if power is already off
-                if (!powerState()) {
-                    break;
+        if (_dev.dev == DEV_SARA_R410) {
+            // If memory issue is present, ensure we don't force a power off too soon
+            // to avoid hitting the 124 day memory housekeeping issue, AT+CPWROFF will
+            // handle this delay automatically, or timeout after 40s.
+            if (_memoryIssuePresent) {
+                MDM_INFO("%s Modem not responsive, waiting up to 30s to power off with PWR_UC...", POWER_OFF_MSG);
+                system_tick_t now = HAL_Timer_Get_Milli_Seconds();
+                if (_timePowerOn == 0) {
+                    // fallback to max timeout of 30s to be safe
+                    _timePowerOn = now;
                 }
-                HAL_GPIO_Write(PWR_UC, 0);
-                // >1.5 seconds on SARA R410M
-                // >1 second on SARA U2
-                // plus a little extra for good luck
-                HAL_Delay_Milliseconds(1600);
-                HAL_GPIO_Write(PWR_UC, 1);
-                break;
-            } else {
-                sendFormated("AT+CPWROFF\r\n");
-                int ret = waitFinalResp(nullptr, nullptr, CPWROFF_TIMEOUT);
-                if (RESP_OK == ret) {
-                    if (_dev.dev == DEV_SARA_R410) {
-                        check_ri = true;
+                // check for timeout (VINT == LOW, Powered on 30s ago, Registered 20s ago)
+                do {
+                    now = HAL_Timer_Get_Milli_Seconds();
+                    // prefer to timeout 20s after registration if we are registered
+                    if (_timeRegistered) {
+                        if (now - _timeRegistered >= 20000UL) {
+                            break;
+                        }
+                    } else if (_timePowerOn && now - _timePowerOn >= 30000UL) {
+                        break;
                     }
-                    break;
-                } else if (RESP_ABORTED == ret) {
-                    MDM_INFO("%s found ABORTED, retrying...", POWER_OFF_MSG);
-                } else {
-                    MDM_INFO("%s timeout, retrying...", POWER_OFF_MSG);
-                }
+                    HAL_Delay_Milliseconds(100); // just wait
+                } while ( powerState() );
+                // reset timers
+                _timeRegistered = 0;
+                _timePowerOn = 0;
             }
         }
-    } else {
-        if (powerState()) {
-            check_ri = true;
-            MDM_INFO("%s modem is on unexpectedly. Turning it off...", POWER_OFF_MSG);
-            // Delay this long period so that the modem is fully on to accept the power-off sequence.
-            HAL_Delay_Milliseconds(12000);
-
+        // Skip power off sequence if power is already off
+        if (powerState() && _dev.dev != DEV_SARA_G350 && _dev.dev != DEV_UNKNOWN) {
```

Comment:
We were attempting power-off through pin 3 times previously, now it's just once.

Reply:
I think we were having one attempt on powering off using pin: https://github.com/particle-iot/device-os/blob/develop/hal/src/electron/modem/mdm_hal.cpp#L1324, right?

Reply:
Good catch, yeah, you are right. :+1: